### ISSUE TITLE
webpack dependency extraction: deprecation warnings on webpack 5

### ISSUE THREAD

**Issue by swissspidy** [WordPress/gutenberg#27984]
**Describe the bug**
When using the dependency extraction webpack plugin with webpack 5, I'm getting  the following warnings:

* `[DEP_WEBPACK_EXTERNALS_FUNCTION_PARAMETERS] DeprecationWarning: The externals-function should be defined like ({context, request}, cb) => { ... }`
* `[DEP_WEBPACK_CHUNK_MODULES_ITERABLE] DeprecationWarning: Chunk.modulesIterable: Use new ChunkGraph API`
* `[DEP_WEBPACK_COMPILATION_ASSETS] DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated.
BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation.`
* `[DEP_WEBPACK_DEPRECATION_ARRAY_TO_SET_PUSH] DeprecationWarning: chunk.files was changed from Array to Set (using Array method 'push' is deprecated)`

Code in question:

https://github.com/WordPress/gutenberg/blob/6dc0b14fe7f983318421dcf3c3f7a2cf02ed3965/packages/dependency-extraction-webpack-plugin/lib/index.js#L48-L49

https://github.com/WordPress/gutenberg/blob/6dc0b14fe7f983318421dcf3c3f7a2cf02ed3965/packages/dependency-extraction-webpack-plugin/lib/index.js#L138-L145

https://github.com/WordPress/gutenberg/blob/6dc0b14fe7f983318421dcf3c3f7a2cf02ed3965/packages/dependency-extraction-webpack-plugin/lib/index.js#L186-L190

**To reproduce**
1. Run `npx webpack ...` orr `node --trace-deprecation node_modules/webpack/bin/webpack.js`

**Expected behavior**

No deprecation warnings should be emitted.

**Screenshots**

**Editor version (please complete the following information):**
- `@wordpress/dependency-extraction-webpack-plugin` version: 2.9.0
- `webpack` version: 5.11.1

**Additional context**
- To report a security issue, please visit the WordPress HackerOne program: https://hackerone.com/wordpress.

**Comment by swissspidy** [comment]
cc @jsnajdr since you worked on #27533

**Comment by gziolo** [comment]
I would expect we drop support for webpack 4 as soon as we migrate everything to webpack 5. However, in practice, the migration might take some time because our first attempt was reverted with https://github.com/WordPress/gutenberg/pull/27974. 😞

**Comment by Levdbas** [comment]
Hi, I am using this plugin for my webpack workflow as well. It seems that this plugin still uses the `emit` hook which causes the `DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated. BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation.` warning. 

As pointed out in https://github.com/webpack/webpack/issues/11425 something like this should be used to support webpack 4 and 5: https://github.com/webpack/webpack/issues/11425#issuecomment-686606318

In my case this causes an issue with the webpack-manifest-plugin where the files generated by this plugin are not in the manifest.

**Comment by jsnajdr** [comment]
> In my case this causes an issue with the `webpack-manifest-plugin` where the files generated by this plugin are not in the manifest.

While the deprecation warnings with webpack 5 are otherwise harmless, and will disappear once we abandon compatibility with webpack 5, the incompatibility with `webpack-manifest-plugin` is something we should solve.

The dependency extraction, and the adding of new assets it does, should happen before `webpack-manifest-plugin` generates a manifest with the list of the assets. Currently, the order is reverse: the `webpack-manifest-plugin`'s `compilation.hooks.processAssets` is called first, the dependency extractions `compiler.emit` hook (a legacy v4 one) is called second.

**Comment by crgeary** [comment]
I've updated the `@wordpress/dependency-extraction-webpack-plugin` package to `3.2.0` (latest) and this issue still exists.

I'm using the following packages (and others):

| package | version |
| --- | --- |
| webpack | 5.46.0 |
| webpack-cli | 4.7.2 |
| @wordpress/dependency-extraction-webpack-plugin | 3.2.0 |

This is the output:

```
(node:41991) [DEP_WEBPACK_EXTERNALS_FUNCTION_PARAMETERS] DeprecationWarning: The externals-function should be defined like ({context, request}, cb) => { ... }
(Use `node --trace-deprecation ...` to show where the warning was created)
(node:41991) [DEP_WEBPACK_CHUNK_MODULES_ITERABLE] DeprecationWarning: Chunk.modulesIterable: Use new ChunkGraph API
(node:41991) [DEP_WEBPACK_COMPILATION_ASSETS] DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated.
BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation.
        Do changes to assets earlier, e. g. in Compilation.hooks.processAssets.
        Make sure to select an appropriate stage from Compilation.PROCESS_ASSETS_STAGE_*.
(node:41991) [DEP_WEBPACK_DEPRECATION_ARRAY_TO_SET_PUSH] DeprecationWarning: chunk.files was changed from Array to Set (using Array method 'push' is deprecated)
```

And this is running on:

- Mac OSX 10.15.7
- Node 16.3.0

**Comment by gziolo** [comment]
Right, the changes landed this week so they weren't included in the last npm publishing round, there is no entry related to this PR in the changelog:

https://unpkg.com/browse/@wordpress/dependency-extraction-webpack-plugin@3.2.0/CHANGELOG.md

It should be published in the next round after Gutenberg 11.2 is released - in two weeks.

**Comment by crgeary** [comment]
I see. I checked the changelog, but as v3.2.0 has been released and that version is not mentioned anywhere the changelog, I assumed the changelog was wrong.

I'll keep an eye out for a future version. Thanks
